### ISSUE TITLE
[GENERAL SUPPORT]: Comments on algorithm

### ISSUE THREAD

**Issue by ad8e** [facebook/Ax#4642]
### Question

I prefer the candidate selection over Vizier's very complex one, but Ax has a few issues that make it only reach parity with my other options, instead of exceeding them:

1. the bounds are given by ranges. This is difficult for us the user. We have to set very loose tolerances for min and max when we're unsure. Protein's method, from Joseph Suarez, accepts a center and scale, without hard cutoffs. This is better for most usecases.
2. What about heteroscedasticity? I'm rarely confident in the scaling of my parameter spaces. I would like some tolerance of these mistakes (configurable tolerance is nice, but optional). This is especially exacerbated when the parameter bounds are ranges. For example, I might have an Attention init scale, whose scale I know is probably similar to my MLP init scale, but I don't know how this parameter's scale relates to my LR scale or beta2 scale. I assume heteroscedasticity is why Ax falls behind HEBO on BBOB1-24 benchmarks.
3. What about NaNs? In our neural networks, the optimal performing configs are on the border of loss spikes. Small loss spikes cause a small increase in final loss, while large loss spikes cause NaNs. So the loss becomes unstable upward at the LR border, and then NaNs once crossing the border. This behavior does not match a GP. With other hyperparameter optimizers, this forces us to throw half our trials away. Inputting a synthetic loss term, such as 1000 - (steps to NaN), does not work.

### Please provide any relevant code snippet if applicable.

```shell

```

### Code of Conduct

- [x] I agree to follow this Ax's Code of Conduct

**Comment by mgrange1998** [comment]
cc @hvarfner questions relating to candidate selection

**Comment by hvarfner** [comment]
Hi @ad8e ,

Thanks for reaching out! Out of curiosity, when you mention issues and parity, do you specifically refer to the performance relative to other methods, or do you have other criteria in mind, too?

1. I think this is fair, and I can see that being the preferred way to go about things in some cases. However, I don't see us changing this in the short term.
2. Do you mean non-stationarity? For heteroskedastic noise, we don't have support out of the box unless the noise level of your problem is observed. 
 
For non-stationary problems (which you seem to be refering to if I'm not mistaken), the output range can vary by large amounts depending on where you are along a dimension. For this, we have the warped SAAS model.

> ...whose scale I know is probably similar to my MLP init scale, but I don't know how this parameter's scale relates to my LR scale or beta2 scale.

If these are all different parameters of the problem you are optimizing, their relative scale does not matter when you specify the problem. Regardless of whether you have warped inputs or not, the GP would determine their relative importance through the lengthscale inference. 

3. Trials which return NaN can be treated in one of two ways, depending on whether
  a. You believe the NaN is transient, and therefore want to try it again, or
  b. The point is bad, and should not be evaluated again.

for a., you mark the trial as "TrialStatus.FAILED", and for b, TrialStatus.ABANDONED. If you mark the trial as FAILED, it is as if it had never happened. Thus, it will not be added to the data. For ABANDONED, we condition on the abandoned input point, which has a repelling effect for future candidate generation.

I would advice against setting extremely large values, since (as you mentioned), this does not resemble a GP. We winsorize large outliers by default, but this would not work for the case where half your data are outliers.


Hopefully this clarifies things, and ask away if you have more questions!

**Comment by ad8e** [comment]
> Out of curiosity, when you mention issues and parity, do you specifically refer to the performance relative to other methods, or do you have other criteria in mind, too?

Other criteria as well. The other major options I know of are Protein, HEBO, and Vizier. (I learned of Ax by talking to a Meta hparam person at NeurIPS.) If I were to pick a hparam optimizer from scratch, I'd try Ax first, because HEBO is plagued by packaging issues and Protein is undocumented. So doing simple things correctly is nice. My codebase already has HEBO implemented, so there's some activation energy to get over the switching cost.

I'm not a serious blackbox optimization researcher, so I use what is convenient.

> Do you mean non-stationarity? For heteroskedastic noise, we don't have support out of the box unless the noise level of your problem is observed.

I was just confused; looking at the HEBO paper, they mean heteroscedasticity as in noise. I only meant some of my axes are more stretched out than others, in an axis-aligned way, which I see you already support. My axes are not sparse, which SAAS appears to aim at. My axes are just poorly scaled.

In ML, a lower bound on noise is observable - the variance in loss between timesteps. But this is a loose lower bound, since loss spikes are either huge or absent depending on luck, so it can severely underestimate noise. These spikes are the main source of heteroscedasticity in noise, when in the optimal region of coefficients.

Thanks for the info on NaNs. That's a great feature. Even though HEBO supports heteroscedastic noise, it does not tolerate NaNs.

Theoretically, modeling these loss spikes would greatly accelerate hparam searches for large ML models, since we already know their nature - they appear when the (effective) LR goes up. But it's not important enough for me to put energy into it. Some models have spikes in the optimal region and some models don't. (Vanilla language transformers do.)

**Comment by hvarfner** [comment]
> I was initially confused; in the HEBO paper, "heteroscedasticity" refers specifically to noise.

I agree! I interpreted the original question (2) as three separate questions, which may have been incorrect. Specifically:

**Do we support heteroskedasticity?**
Unless the noise level is known, we do not. If the heteroskedasticity is significant (and you know it is due to noise, rather than rapid changes in the underlying function), you may encounter issues with optimization efficiency. You could apply a power transform (as in HEBO), but personally, I do not see a clear advantage in using this approach to address heteroskedasticity.

**Do hyperparameters need to be scaled similarly, so that each is roughly equally important?**
No. The surrogate model will naturally infer the relative importance of each hyperparameter and adjust the Bayesian Optimization process accordingly. We also have analyses and visualizations to support this.

**For a hyperparameter with a broad range, can we support different function behavior in different parts of that range?**
Yes, through input warpings (as in HEBO). This can be configured in the ModelConfig if you believe it would benefit your problem.


> In theory, modeling these loss spikes could greatly accelerate hyperparameter searches for large ML models, since we already understand their nature—they occur when the (effective) learning rate increases. However, it is not a priority for me at the moment.

Handling these spikes (or NaNs) is certainly a significant challenge, and we are continually exploring the best ways to address it. Currently, we winsorize large outliers and apply the aforementioned conditioning for abandoned trials.
